### PR TITLE
slicing across cubes

### PR DESCRIPTION
**Pull request by abit2**
slicing across cubes. @DanRyanIrish

### CODE REVIEW

**Comment by abit2** [comment]
corrected the failing test pixelize_slice (no segmentation fault :P). Added test and made some changes/added lines in code. Made comments in code this time. :) @DanRyanIrish

File: sunpycube/cube/cube_utils.py [hunk 1]
```diff
@@ -613,10 +633,10 @@ def index_sequence_as_cube(cubesequence, item):
...
         sequence_index, cube_index = _convert_cube_like_slice_to_sequence_slices(
             item_list[cubesequence.common_axis], cumul_cube_lengths)
-        if sequence_index.start != sequence_index.stop:
-            raise ValueError("Slicing across multiple cubes not yet supported.")
-        else:
-            sequence_index = sequence_index.start
+        # if sequence_index.start != sequence_index.stop:
+        #     raise ValueError("Slicing across multiple cubes not yet supported.")
+        # else:
+        #     sequence_index = sequence_index.start
```
Comment (DanRyanIrish) [inline sunpycube/cube/cube_utils.py]: Delete these commented out lines as slicing across cubes should now be supported.

File: sunpycube/cube/cube_utils.py [hunk 2]
```diff
@@ -519,16 +517,38 @@ def get_cube_from_sequence(cubesequence, item):
     """
     if isinstance(item, int):
         result = cubesequence.data[item]
-
     if isinstance(item, slice):
-        data = cubesequence.data[item]
-        meta = cubesequence.meta
-        result = cubesequence._new_instance(data, meta)
-
+        data = cubesequence.data[item.start:item.stop]
```
Comment (DanRyanIrish) [inline sunpycube/cube/cube_utils.py]: Is there a reason you don't you just index with the slice object itself?  That way a user could use the slice step to get e.g. every 2nd cube if they want.

File: sunpycube/cube/cube_utils.py [hunk 3]
```diff
@@ -636,19 +656,53 @@ def _convert_cube_like_index_to_sequence_indices(cube_like_index, cumul_cube_len
         cube_index = cube_like_index
     else:
         sequence_index = np.where(cumul_cube_lengths <= cube_like_index)[0][-1]
-        cube_index = cube_like_index - cumul_cube_lengths[sequence_index]
-        # sequence_index should be plus one as the sequence_index earlier is previous index.
-        sequence_index += 1
+        # if the cube is out of range then return the last index
+        if cube_like_index > cumul_cube_lengths[-1] - 1:
+            cube_index = cumul_cube_lengths[0] - 1
```
Comment (DanRyanIrish) [inline sunpycube/cube/cube_utils.py:657]: If the user enters an exposure number which is out of range of the number of exposures in the CubeSequence, I think an error should be raised rather.

File: sunpycube/cube/datacube.py [hunk 4]
```diff
@@ -554,20 +555,16 @@ class CubeSequence(object):
...
-        """
-        Instantiate a new instance of this class using given data.
-        """
-        return cls(data_list, meta=meta)
+        try:
+            self.shape = tuple([len(data_list)] + list(data_list[0].shape))
+        except AttributeError as err:
+            warnings.warn("AttributeError " + str(err))
+        else:
+            self.shape = tuple([len(data_list)])
```
Comment (DanRyanIrish) [inline sunpycube/cube/datacube.py]: Is the purpose of this try...except statement to make CubeSequence work for non-Cubes/arrays?

Comment (abit2) [reply sunpycube/cube/datacube.py]: Yes!! lightcurve does not have shape attribute

File: sunpycube/cube/cube_utils.py (hunk 3 above)
Comment (abit2) [reply sunpycube/cube/cube_utils.py:657]:
```python
In [4]: a = np.array([1,2,3,4,5])

In [5]: a[0:100]
Out[5]: array([1, 2, 3, 4, 5])

```

@DanRyanIrish array behaves in this way so thought it should be same.

Comment (DanRyanIrish) [reply sunpycube/cube/cube_utils.py:657]: Wow!  I did not realise that!  In that case you've done the right thing :)

File: sunpycube/cube/cube_utils.py (hunk 2 above)
Comment (DanRyanIrish) [reply sunpycube/cube/cube_utils.py]: I see you indexed with just the slice below.  Is there a reason you haven't done that here too?

File: sunpycube/cube/tests/test_cube_utils.py [hunk 5]
```diff
@@ -115,13 +115,39 @@ def test_convert_slice():
...
+
+
+@pytest.mark.parametrize("test_input,expected", [
+    (cu._convert_cube_like_index_to_sequence_indices(5, np.array([8, 16, 24, 32])), (0, 5)),
+    (cu._convert_cube_like_index_to_sequence_indices(8, np.array([8, 16, 24, 32])), (1, 0)),
+    (cu._convert_cube_like_index_to_sequence_indices(20, np.array([8, 16, 24, 32])), (2, 4)),
+    (cu._convert_cube_like_index_to_sequence_indices(50, np.array([8, 16, 24, 32])), (3, 7)),
+])
+def test_convert_cube_like_index_to_sequence_indices(test_input, expected):
+    assert test_input == expected
```
Comment (DanRyanIrish) [inline sunpycube/cube/tests/test_cube_utils.py:137]: Does this and other tests also check whether it errors in the right way?

File: sunpycube/cube/tests/test_cube_utils.py [hunk 6]
```diff
@@ -115,13 +115,39 @@ def test_convert_slice():
...
+
+
+@pytest.mark.parametrize("test_input,expected", [
+    (cu._convert_cube_like_slice_to_sequence_slices(slice(2, 5),
```
Comment (DanRyanIrish) [inline sunpycube/cube/tests/test_cube_utils.py]: Isn't this a circular test?  Using the function to give the answer that you want to test? If so, wont that mean it wont fail even if the function is wrong?  Would it be better better to have explicit expected inputs and outputs?

File: sunpycube/cube/cube_utils.py (hunk 2 above)
Comment (abit2) [reply sunpycube/cube/cube_utils.py]: no it is my mistake will be changing this :)

File: sunpycube/cube/tests/test_cube_utils.py [hunk 7]
```diff
@@ -115,13 +115,39 @@ def test_convert_slice():
...
+                                                    np.array([8, 16, 24, 32])), (slice(0, 1), slice(2, 5))),
+    (cu._convert_cube_like_slice_to_sequence_slices(slice(5, 15), np.array(
+        [8, 16, 24, 32])), (slice(0, 2), [slice(5, 8), slice(0, 7)])),
+    (cu._convert_cube_like_slice_to_sequence_slices(slice(5, 16),
+                                                    np.array([8, 16, 24, 32])), (slice(0, 2), [slice(5, 8)])),
```
Comment (DanRyanIrish) [inline sunpycube/cube/tests/test_cube_utils.py]: Should the end of this be ```[slice(5, 8), slice(0, 8)]``` ?

File: sunpycube/cube/tests/test_cube_utils.py [hunk 8]
```diff
@@ -115,13 +115,39 @@ def test_convert_slice():
...
+    (cu._convert_cube_like_slice_to_sequence_slices(slice(5, 23), np.array(
+        [8, 16, 24, 32])), (slice(0, 3), [slice(5, 8), slice(0, 7)])),
+    (cu._convert_cube_like_slice_to_sequence_slices(slice(5, 100),
+                                                    np.array([8, 16, 24, 32])), (slice(0, 4), [slice(5, 8)])),
```
Comment (DanRyanIrish) [inline sunpycube/cube/tests/test_cube_utils.py:150]: Should the output of this be ```(slice(0, 6), [slice(5, 8), slice(0,4)])```?

Comment (abit2) [reply sunpycube/cube/tests/test_cube_utils.py:150]: No the tuple[0] is the sequence_slice which will be from (0, 4) and the tuple[1] is the list which consist of the slices of 1st and last cube which if it needs to be sliced.

Comment (DanRyanIrish) [reply sunpycube/cube/tests/test_cube_utils.py:150]: OK.  It's good to have that confirmed.  But here, you want to get the 5th to the 99th (inclusive) exposures, right?  There are 16 raster positions in each cube, right?  So the first exposure of the 7th cube brings you to the 96th exposure.  (This is all 0-based counting.)  So to get to the 100th exposure you need slice(0,4) of the 7th cube.  So in fact you need ```(slice(0, 7), [slice(5, 8), slice(0,4)])```?

File: sunpycube/cube/tests/test_cubesequence.py [hunk 9]
```diff
@@ -58,6 +60,22 @@ def test_slice_first_index_sequence(test_input, expected):
...
     (seq[0:].shape[0], 4),
+    (seq[slice(0, 2)].shape[0], 2),
+    (seq[slice(0, 3)].shape[0], 3),
 ])
 def test_slice_first_index_sequence(test_input, expected):
     assert test_input == expected
+
+
+@pytest.mark.parametrize("test_input,expected", [
+    (seq.index_as_cube(0), Spectrum),
```
Comment (DanRyanIrish) [inline sunpycube/cube/tests/test_cubesequence.py]: Should this not also be a GenericMap?

File: sunpycube/cube/tests/test_cube_utils.py (hunk 8 above)
Comment (DanRyanIrish) [reply sunpycube/cube/tests/test_cube_utils.py:150]: Fixed.

**Review by DanRyanIrish — APPROVED** [review]

**Pull request merged**